raiden/tests/unit/test_utils.py

```diff
@@ -7,3 +12,62 @@ def test_privtopub():
               '705f70c7554b26e82b90d2d1bbbaf711b10c6c8b807077f4070200a8fb4c6b771')
 
     assert pubkey == privtopub(privkey).hex()
+
+
+def test_patched_gevent_exception_handling(with_configure_gevent):
+    with pytest.raises(_ExceptionInGreenlet):  # exception is bubbled up
+        greenlet = gevent.spawn(_greenlet)
+        greenlet.join()
+
+    with pytest.raises(_ExceptionInGreenlet):
+        greenlet = gevent.spawn_later(0.1, _greenlet)
+        greenlet.join()
+
+    try:
+        greenlet1 = gevent.spawn(_greenlet)
+        greenlet2 = gevent.spawn_later(0.1, _greenlet)
+        greenlet3 = gevent.spawn_later(0.1, _greenlet)
+        greenlet1.link_exception_safe(_swallow_exception)
+        greenlet2.rawlink_safe(_swallow_exception)
+        greenlet3.link_safe(_swallow_exception)
+        greenlet1.join()
+        greenlet2.join()
+        greenlet3.join()
+    except Exception:
+        assert False, 'This should not be propagated out of the greenlet'
+
+    with pytest.raises(UnhandledExceptionInGreenlet, message='_ExceptionInGreenlet (msg)'):
+        greenlet = gevent.spawn_later(0.1, _greenlet)
+        greenlet.link_exception_safe(_reraise_exception)
+        greenlet.join()
+
+    with pytest.raises(UnhandledExceptionInGreenlet, message='_ExceptionInGreenlet (msg)'):
+        greenlet = gevent.spawn(_greenlet)
+        greenlet.rawlink_safe(_reraise_exception)
+        greenlet.join()
+
+    with pytest.raises(UnhandledExceptionInGreenlet, message='_ExceptionInGreenlet (msg)'):
+        greenlet = gevent.spawn(_greenlet)
+        greenlet.link_safe(_reraise_exception)
+        greenlet.join()
+
+
+class _ExceptionInGreenlet(Exception):
+    pass
+
+
+def _greenlet():
```

Comment:
Please give this a better name. For a while I thought it was something from gevent. `_failing_greenlet` or `_exception_raising_greenlet` is one example of a better name